kernel/src/arch/x86_64/rackscale/controller_state.rs

```diff
@@ -4,121 +4,104 @@
 use alloc::boxed::Box;
 use alloc::sync::Arc;
 use alloc::vec::Vec;
-use hashbrown::HashMap;
+
+use arrayvec::ArrayVec;
 use lazy_static::lazy_static;
 use spin::Mutex;
-use static_assertions as sa;
 
 use kpi::system::{CpuThread, MachineId};
 
-use crate::arch::rackscale::dcm::DCMNodeId;
-use crate::memory::mcache::MCache;
-use crate::memory::LARGE_PAGE_SIZE;
+use crate::arch::rackscale::FrameCacheBase;
+use crate::arch::MAX_MACHINES;
+use crate::memory::backends::MemManager;
+use crate::memory::shmem_affinity::{local_shmem_affinity, mid_to_shmem_affinity};
+use crate::memory::{mcache::MCache, LARGE_PAGE_SIZE};
 use crate::transport::shmem::get_affinity_shmem;
 
+/// TODO(rackscale): think about how we should constrain this?
 /// Global state about the local rackscale client
 lazy_static! {
-    pub(crate) static ref CONTROLLER_AFFINITY_SHMEM: Arc<Mutex<Box<FrameCacheShmem>>> =
-        Arc::new(Mutex::new(
-            get_affinity_shmem()
-                .get_shmem_manager(0)
-                .expect("Failed to fetch shmem manager for controller shmem.")
-        ));
-}
+    pub(crate) static ref CONTROLLER_SHMEM_CACHES: Arc<Mutex<ArrayVec<Box<dyn MemManager + Send>, MAX_MACHINES>>> = {
```

Comment:
Is this accessed a lot? if so might be better to do Arc<ArrayVec<Mutex<>>> instead of Arc<Mutex<ArrayVec<>>>

I guess it's controller state so maybe it won't matter since it's single core?

Reply:
Yes, that was my thinking. I can change that now in this PR or wait until later - no strong preference on my part

Reply:
Now need to change now I think